### PR TITLE
Slider: Fix to restrict handle values when set programmatically.

### PR DESCRIPTION
**Pull request by dekajp**
Fixes #3762 - Slider: handles not restricted properly when set programmatically

### CODE REVIEW

File: ui/slider.js [hunk 1]
```diff
@@ -157,6 +158,48 @@ return $.widget( "ui.slider", $.ui.mouse, {
 		}
 	},
 
+	_cleanValues: function( index, newValue ) {
+		var i = 0, values ;
```
Comment (scottgonzalez) [inline ui/slider.js]: Uninitialized variables go at the top:

```js
var values,
    i = 0;
```

File: ui/slider.js [hunk 2]
```diff
@@ -157,6 +158,48 @@ return $.widget( "ui.slider", $.ui.mouse, {
...
+			if ( index === 0 && newValue > values[ index + 1 ] ) {
+				return values[ index + 1 ];
+			}
+			if ( index === 1 && newValue < values[ index - 1 ] ) {
+				return values[ index - 1 ];
+			}
+			return newValue;
+		}
+
+		if ( arguments.length ) {
```
Comment (scottgonzalez) [inline ui/slider.js]: Am I missing something here? There are multiple calls to this method with no parameters, but we do nothing in that case?

File: ui/slider.js [hunk 3]
```diff
@@ -157,6 +158,48 @@ return $.widget( "ui.slider", $.ui.mouse, {
...
+
+		if ( arguments.length > 1 ) {
+			values = this.options.values;
+			if ( index === 0 && newValue < this._valueMin() ){
+				newValue = this._valueMin();
+			}
+			if ( index === 1 && newValue > this._valueMax() ){
+				newValue = this._valueMax();
+			}
+			if ( index === 0 && newValue > values[ index + 1 ] ) {
```
Comment (scottgonzalez) [inline ui/slider.js]: This logic should only apply with ranges.

**Comment by scottgonzalez** [comment]
There are some spacing issues and I didn't do a full review because it looks like this still needs some changes to how it works.

File: ui/slider.js (hunk 2 above)
Comment (dekajp) [inline ui/slider.js]: My bad , I missed them.

**Comment by jzaefferer** [comment]
@dekajp please comment on your PRs after you push an update, since the push itself has no notification.

@scottgonzalez there's a new commit since yesterday

File: ui/slider.js [hunk 4]
```diff
@@ -422,6 +465,7 @@ return $.widget( "ui.slider", $.ui.mouse, {
 				this._refreshValue();
 			} else {
 				if ( this.options.values && this.options.values.length ) {
+					this.options.values = this._cleanValues( this.options.values );
```
Comment (scottgonzalez) [inline ui/slider.js]: Why are the values being cleaned during a get?

File: ui/slider.js [hunk 5]
```diff
@@ -157,6 +157,48 @@ return $.widget( "ui.slider", $.ui.mouse, {
 		}
 	},
 
+	_cleanValues: function( index, newValue ) {
+		var values,
+			i = 0 ;
+
+		if ( $.isArray( arguments[ 0 ] ) ) {
+			values = arguments[ 0 ] || this.options.values;
```
Comment (scottgonzalez) [inline ui/slider.js]: How would this ever be `this.options.values` since `arguments[ 0 ]` is guaranteed to be an array?

File: ui/slider.js [hunk 6]
```diff
@@ -157,6 +157,48 @@ return $.widget( "ui.slider", $.ui.mouse, {
 		}
 	},
 
+	_cleanValues: function( index, newValue ) {
+		var values,
+			i = 0 ;
```
Comment (scottgonzalez) [inline ui/slider.js]: extraneous space

File: ui/slider.js [hunk 7]
```diff
@@ -157,6 +157,48 @@ return $.widget( "ui.slider", $.ui.mouse, {
...
+
+		if ( $.isArray( arguments[ 0 ] ) ) {
+			values = arguments[ 0 ] || this.options.values;
+			if ( values ) {
```
Comment (scottgonzalez) [inline ui/slider.js]: Why does this check exist?

File: ui/slider.js [hunk 8]
```diff
@@ -157,6 +157,48 @@ return $.widget( "ui.slider", $.ui.mouse, {
...
+				for ( ; i < values.length; i++ ){
```
Comment (scottgonzalez) [inline ui/slider.js]: Missing space before the brace.

File: ui/slider.js [hunk 9]
```diff
@@ -157,6 +157,48 @@ return $.widget( "ui.slider", $.ui.mouse, {
...
+					if ( values[ i ] < this._valueMin() ) {
+						values[ i ] = this._valueMin();
+					}
+
+					if ( values[ i ]  > this._valueMax() ) {
+						values[ i ] = this._valueMax();
+					}
+				}
+				if ( values[ 0 ] > values[ 1 ] ) {
```
Comment (scottgonzalez) [inline ui/slider.js]: This is wrong for non-range sliders.

File: ui/slider.js [hunk 10]
```diff
@@ -157,6 +157,48 @@ return $.widget( "ui.slider", $.ui.mouse, {
...
+			}
+		}
+
+		if ( arguments.length > 1 ) {
+			values = this.options.values;
+			if ( index === 0 && newValue < this._valueMin() ){
+				newValue = this._valueMin();
+			}else if ( index === 1 && newValue > this._valueMax() ){
+				newValue = this._valueMax();
+			}
```
Comment (scottgonzalez) [inline ui/slider.js]: Why is this only half of the min/max logic?

File: ui/slider.js [hunk 11]
```diff
@@ -157,6 +157,48 @@ return $.widget( "ui.slider", $.ui.mouse, {
...
+
+			if ( this.options.range === true && values.length === 2 ) {
```
Comment (scottgonzalez) [inline ui/slider.js]: The length check isn't necessary. Range sliders must have exactly two values.

File: ui/slider.js [hunk 12]
```diff
@@ -157,6 +157,48 @@ return $.widget( "ui.slider", $.ui.mouse, {
...
+				if ( index === 0 && newValue > values[ index + 1 ] ) {
+					newValue = values[ index + 1 ];
+				}else if ( index === 1 && newValue < values[ index - 1 ] ) {
+					newValue = values[ index - 1 ];
+				}
```
Comment (scottgonzalez) [inline ui/slider.js]: Why the `index +/- 1` logic? You already have the indices hard-coded anyway out of necessity.

**Comment by scottgonzalez** [comment]
Maybe we should start with just a set of tests so that we can agree the correct behavior is being tested.

File: ui/slider.js (hunk 4 above)
Comment (dekajp) [inline ui/slider.js]: I will change this and move `_cleanValues` to create

**Comment by dekajp** [comment]
do you want me to add/modify the unit test cases ?

File: ui/slider.js (hunk 10 above)
Comment (dekajp) [inline ui/slider.js]: If `range = true`   `min <= X0 <= X1` and  `X0 <= X1<= Max`

**Comment by dekajp** [comment]
From http://bugs.jqueryui.com/ticket/3762#comment:31
1. Restricting the value of a handle based on other `handles only applies with range: true` and therefore also only applies when exactly two handles exist. 
2. When setting both values at the same time, either via .values( array ), .option( "values", array ), or initialization,
       1.  the first handle can be any value between min and max and
       2.  the second handle must be between the first handle and max. 
3. When setting a single handle, it is restricted by the min, max, and existing value of the other handle.

**Comment by dekajp** [comment]
@scottgonzalez  , i will send a new PR  on this.

File: tests/unit/slider/slider_methods.js [hunk 13]
```diff
@@ -100,8 +100,90 @@ test( "value", function() {
...
+	element.slider( "values", [ 0, 200 ] );
+	deepEqual( element.slider( "values" ), [ 10, 100 ], "Values (array) - restricted min and max" );
+
+	element.slider( "values", 0, 50 );
+	equal( element.slider( "values", 0 ), 50, "Values (index,value) - set first handle" );
+
+	element.slider( "values", 1, 75 );
+	equal( element.slider( "values", 1 ), 75, "Values (index,value) - set second handle" );
+
+	element.slider( "values", [ 50, 75 ] );
```
Comment (scottgonzalez) [inline tests/unit/slider/slider_methods.js]: This is effectively a no-op. Why is it here?

File: tests/unit/slider/slider_methods.js [hunk 14]
```diff
@@ -100,8 +100,90 @@ test( "value", function() {
...
+	element.slider( "values", 0, 5 );
+	equal( element.slider( "values", 0 ), 10, "Values (index,value) - 1st handle restricted properly, against min" );
+
+	element.slider( "values", 1, 5 );
+	equal( element.slider( "values", 0 ), 10, "Values (index,value) - 2nd handle restricted properly, against min" );
+
+	element.slider( "values", [ 50, 75 ] );
+	element.slider( "values", 0, 110 );
+	equal( element.slider( "values", 0 ), 75, "Values (index,value) - 1st handle restricted properly, against max" );
```
Comment (scottgonzalez) [inline tests/unit/slider/slider_methods.js]: It's restricted by the second handle, not the max.

File: tests/unit/slider/slider_methods.js [hunk 15]
```diff
@@ -100,8 +100,90 @@ test( "value", function() {
...
+
+	element.slider( "values", 1, 110 );
+	equal( element.slider( "values", 1 ), 100, "Values (index,value) - 2nd handle restricted properly, against max" );
+
+	element.slider( "values", [ 50, 75 ] );
+	element.slider( "values", 0, 90 );
+	equal( element.slider( "values", 0 ), 75, "Values (index,value) - 1st handle restricted properly, against 2nd handle" );
```
Comment (scottgonzalez) [inline tests/unit/slider/slider_methods.js]: This is already tested in L134.

File: tests/unit/slider/slider_methods.js [hunk 16]
```diff
@@ -100,8 +100,90 @@ test( "value", function() {
...
+	element.slider( "values", [1.5, 1.3] );
+	deepEqual( element.slider( "values" ), [ 1.5, 1.5 ], "Values (array) - Floating Point - uneven step - 2nd Handle restricted" );
+
+	element.slider( "values", [1.5, 1.5] );
+	element.slider( "values", 1, 1.3 );
+	deepEqual( element.slider( "values" ), [ 1.5, 1.5 ], "Values (index,value) - Floating Point - uneven step - 2nd Handle restricted" );
+
+	element.slider( "values", [1.5, 1.5] );
+	element.slider( "values", 0, 1.6 );
+	deepEqual( element.slider( "values" ), [ 1.5, 1.5 ], "Values (index,value) - Floating Point - uneven step - 1st Handle restricted" );
```
Comment (scottgonzalez) [inline tests/unit/slider/slider_methods.js]: Why are we doing a whole slew of floating point tests for boundary restrictions?

**Comment by scottgonzalez** [comment]
I think the tests would be easier to follow if they were grouped by which handle is being tested. For example, testing the individual handle settings can be simplified to:

* Set the first handle below min, verify it goes to min.
* Set the first handle above the second handle, verify it goes to the second handle.
* Set the second handle above the max, verify it goes to the max.
* Set the second handle below the first handle, verify it goes to the first handle.

There's no need to reset the values in between because every individual assertion can be properly performed with a single value change. You can test the multi-value setter in a similar grouping.

File: tests/unit/slider/slider_methods.js (hunk 15 above)
Comment (dekajp) [inline tests/unit/slider/slider_methods.js]: element.slider( "values", 0, 110 );  ` Max = 100 `

element.slider( "values", 0, 90 ); ` Less than max but greater than 2nd handle`

Comment (scottgonzalez) [inline tests/unit/slider/slider_methods.js]: But in both cases, the handle has to be restricted by the second handle.

Comment (dekajp) [inline tests/unit/slider/slider_methods.js]: Alright

File: tests/unit/slider/slider_methods.js (hunk 16 above)
Comment (dekajp) [inline tests/unit/slider/slider_methods.js]: Well just making sure floating points are covered

Comment (scottgonzalez) [inline tests/unit/slider/slider_methods.js]: That's not necessary.

**Comment by dekajp** [comment]
@scottgonzalez  , I rebased my files. Now github is not showing here , asking me to send PR
Also `grunt` is giving lint error in html in master branch . 

https://github.com/dekajp/jquery-ui/commit/ec53b6d0ed7aaa5412fa3475f77a04ab1de0a4c7

**Comment by scottgonzalez** [comment]
It looks like the branch names are different (the casing on the v2).

**Comment by scottgonzalez** [comment]
@dekajp What's the status of this?

**Comment by dekajp** [comment]
@scottgonzalez   last 3 weeks i was busy with relocation. i think this weekend i will be able to get back on this

**Comment by scottgonzalez** [comment]
Thanks. Hope the move went well.

**Comment by dekajp** [comment]
@scottgonzalez  yeah move went all , everyone happy in family :) with more space .  I rebased commits .

File: tests/unit/slider/methods.js [hunk 17]
```diff
@@ -109,8 +109,54 @@ test( "value", function() {
...
+
+	element.slider( "values", 1, 75 );
+	equal( element.slider( "values", 1 ), 75, "Values (index,value) - set second handle" );
+
+	
+	element.slider( "values", 0, 5 );
+	equal( element.slider( "values", 0 ), 10, "Values (index,value) - 1st handle restricted properly, against min" );
+
+	element.slider( "values", 0, 90 );
+	equal( element.slider( "values", 0 ), element.slider( "values", 1 ), "Values (index,value) - 1st handle restricted properly, against 2nd handle" );
```
Comment (scottgonzalez) [inline tests/unit/slider/methods.js]: I understand that you're trying to show that the handles are matching, but I'd prefer an explicit value for the expected value.

File: tests/unit/slider/methods.js [hunk 18]
```diff
@@ -109,8 +109,54 @@ test( "value", function() {
...
+
+	element.slider( "values", 1, 110 );
+	equal( element.slider( "values", 1 ), 100, "Values (index,value) - 2nd handle restricted properly, against max" );
+
+	element.slider( "values", 1, 40 );
+	equal( element.slider( "values", 1 ), element.slider( "values", 0 ), "Values (index,value) - 2nd handle restricted properly, against 1st handle" );
```
Comment (scottgonzalez) [inline tests/unit/slider/methods.js]: Same change here, use an explicit value.

File: tests/unit/slider/methods.js [hunk 19]
```diff
@@ -109,8 +109,54 @@ test( "value", function() {
...
+	
+	element.slider( "values", [ 50, 75 ] );
```
Comment (scottgonzalez) [inline tests/unit/slider/methods.js]: This is still effectively a no-op and should be removed.

File: tests/unit/slider/methods.js [hunk 20]
```diff
@@ -109,8 +109,54 @@ test( "value", function() {
...
+
+	element.slider( "values", [ 0, 75 ] );
+	deepEqual( element.slider( "values" ), [ 10, 75 ], "Values (array) - Setting both handles - against  min" );
```
Comment (scottgonzalez) [inline tests/unit/slider/methods.js]: You've already asserted this on L125.

File: tests/unit/slider/methods.js [hunk 21]
```diff
@@ -109,8 +109,54 @@ test( "value", function() {
...
+
+	element.slider( "values", [ 0, 90 ] );
+	deepEqual( element.slider( "values" ), [ 10, 90 ], "Values (array) - 1st handle restricted properly, against 2nd handle" );
```
Comment (scottgonzalez) [inline tests/unit/slider/methods.js]: This is exactly the same as the previous assertion, and the message is wrong.

File: tests/unit/slider/methods.js [hunk 22]
```diff
@@ -109,8 +109,54 @@ test( "value", function() {
...
+
+	element.slider( "values", 1, 110 );
+	deepEqual( element.slider( "values" ), [ 10, 100 ], "Values (array) - 2nd handle restricted properly, against max" );
```
Comment (scottgonzalez) [inline tests/unit/slider/methods.js]: You've already asserted this on L125.

File: tests/unit/slider/methods.js [hunk 23]
```diff
@@ -109,8 +109,54 @@ test( "value", function() {
...
+	
+	element.slider( "values", [ 75, 50 ] );
+	deepEqual( element.slider( "values" ), [ 75, 75 ], "Values (array) - Setting both handles - 2nd Handle resets to 1st" );
```
Comment (scottgonzalez) [inline tests/unit/slider/methods.js]: I wouldn't use the word "resets" here; it's being restricted.

File: tests/unit/slider/methods.js [hunk 24]
```diff
@@ -109,8 +109,54 @@ test( "value", function() {
...
+});
```
Comment (scottgonzalez) [inline tests/unit/slider/methods.js]: Let's add some assertions for non-range sliders as well, to verify the handle restrictions aren't applied.

File: tests/unit/slider/methods.js [hunk 25]
```diff
@@ -109,8 +109,54 @@ test( "value", function() {
...
+	expect( 12 );
+	
+	var element = $( "<div></div>" ).slider({
+		range: true,
+		min: 10,
+		max: 100,
+		step: 1
+	});
+
+	element.slider( "values", [ 20, 90 ] );
```
Comment (scottgonzalez) [inline tests/unit/slider/methods.js]: Set the value in the initialization.

File: tests/unit/slider/options.js [hunk 26]
```diff
@@ -264,6 +265,26 @@ test( "values", function() {
...
+	});
+	deepEqual( element.slider( "values" ), [ 15, 15 ], "handles restricted properly when set programmatically - 2nd handle reset" );
+
+	element.slider( "option", "values", [ 0, 200 ] );
+	deepEqual( element.slider( "values" ), [ 10, 100 ], "option values - restricted min and max" );
+
+	element.slider( "option", "values", [ 15, 1 ] );
+	deepEqual( element.slider( "values" ), [ 15, 15 ], "option values - restricted 2nd Handle" );
+
+	element.slider( "option", "values", [ 15, 15 ] );
```
Comment (scottgonzalez) [inline tests/unit/slider/options.js]: This is effectively a no-op and should be removed.

File: ui/slider.js [hunk 27]
```diff
@@ -160,6 +161,41 @@ return $.widget( "ui.slider", $.ui.mouse, {
...
+
+				if ( !arguments.length ){
+					this.options.values = values;
+				}
+
+				return values;
+			}
+		} else {
+			if ( $.isArray( arguments[ 0 ] ) ) {
+				return arguments[0];
```
Comment (scottgonzalez) [inline ui/slider.js]: Bad spacing.

File: ui/slider.js [hunk 28]
```diff
@@ -160,6 +161,41 @@ return $.widget( "ui.slider", $.ui.mouse, {
...
 
+	_cleanValues: function( index, newValue ) {
+		var values;
+
+		if ( this.options.range === true ){
+			if ( arguments.length === 2 ) {
+				values = this.options.values;
+				if ( index === 0 && newValue > values[ 1 ] ) {
+					newValue = values[ 1 ];
+				}else if ( index === 1 && newValue < values[ 0 ] ) {
```
Comment (scottgonzalez) [inline ui/slider.js]: bad spacing

File: ui/slider.js [hunk 29]
```diff
@@ -160,6 +161,41 @@ return $.widget( "ui.slider", $.ui.mouse, {
...
+					newValue = values[ 0 ];
+				}
+				return newValue;
+			}
+
+			values = arguments[ 0 ] || this.options.values;
+			if ( values ) {
```
Comment (scottgonzalez) [inline ui/slider.js]: Why does this conditional exist?

File: ui/slider.js [hunk 30]
```diff
@@ -160,6 +161,41 @@ return $.widget( "ui.slider", $.ui.mouse, {
...
+				if ( values[ 0 ] > values[ 1 ] ) {
+					values[ 1 ] = values[ 0 ];
+				}
+
+				if ( !arguments.length ){
```
Comment (scottgonzalez) [inline ui/slider.js]: Bad spacing.

File: ui/slider.js [hunk 31]
```diff
@@ -160,6 +161,41 @@ return $.widget( "ui.slider", $.ui.mouse, {
...
+					this.options.values = values;
```
Comment (scottgonzalez) [inline ui/slider.js]: It seems odd that this method is setting internal state in this one case.

Comment (scottgonzalez) [inline ui/slider.js]: I think it's just the way this method is organized that makes it seem odd.

File: tests/unit/slider/methods.js (hunk 19 above)
Comment (dekajp) [inline tests/unit/slider/methods.js]: It is hard to follow so many test cases so i reset values , for easy readability.

Comment (scottgonzalez) [inline tests/unit/slider/methods.js]: But the next line of code resets values as well. The code isn't any easier to follow with this additional step.

File: tests/unit/slider/methods.js (hunk 25 above)
Comment (jzaefferer) [inline tests/unit/slider/methods.js]: All of these were addressed by #1573, which landed in 4c5cd2018d88392871cfc93d3fb6f71a0bc510ff

**Comment by jzaefferer** [comment]
@dekajp would you be available to rebase this and update the implementation?

**Comment by dekajp** [comment]
@jzaefferer i rebased by commit. I will look into update the implementation

**Comment by sameer05k21a0440** [comment]
I have dynamic multiple input and dynamic slider handle so when on change input text value need to set slider value but I can't set slider handle check below code .

```
 $(this).parents("tr").find('input[name="assesmentScore"]').change( function (event) {
	   					 var textValue = $(this).parents("tr").find('input[name="assesmentScore"]').val();
$("#bonusScore", $(this)).slider('value',textValue);    //here need to add dynamic slide how ? 
	   					handle.text(textValue); //Can't set corresponding text
});
```

Note :Please let me if needed full code.
Thanks

**Comment by dmethvin** [comment]
Closing as a stale branch.
